batched/dense/impl/KokkosBatched_Laswp_Serial_Internal.hpp

```diff
@@ -0,0 +1,116 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOSBATCHED_LASWP_SERIAL_INTERNAL_HPP_
+#define KOKKOSBATCHED_LASWP_SERIAL_INTERNAL_HPP_
+
+/// \author Yuuichi Asahi (yuuichi.asahi@cea.fr)
+
+#include "KokkosBatched_Util.hpp"
+
+namespace KokkosBatched {
+namespace Impl {
+
+///
+/// Serial Internal Impl
+/// ========================
+
+///
+//// Forward pivot apply
+///
+
+struct SerialLaswpVectorForwardInternal {
+  template <typename IntType, typename ValueType>
+  KOKKOS_INLINE_FUNCTION static int invoke(const int plen, const IntType *KOKKOS_RESTRICT p, const int ps0,
+                                           /* */ ValueType *KOKKOS_RESTRICT A, const int as0) {
+    for (int i = 0; i < plen; ++i) {
+      const int piv = p[i * ps0];
+      if (piv != i) {
+        const int idx_i = i * as0, idx_p = piv * as0;
+        const ValueType tmp = A[idx_i];
+        A[idx_i]            = A[idx_p];
+        A[idx_p]            = tmp;
+      }
+    }
+    return 0;
+  }
+};
+
+struct SerialLaswpMatrixForwardInternal {
+  template <typename IntType, typename ValueType>
+  KOKKOS_INLINE_FUNCTION static int invoke(const int n, const int plen, const IntType *KOKKOS_RESTRICT p, const int ps0,
+                                           /* */ ValueType *KOKKOS_RESTRICT A, const int as0, const int as1) {
+    for (int j = 0; j < n; j++) {
```

Comment:
Are we assuming layout left? Otherwise it would make sense for a layout right to inverse these two loops.

Reply:
Actually, the expected layout is `LayoutStrided` since `A` is supposed to be a subview of batches of `A`. Based on your comments, I have changed the loop order by comparing the strides of `A` (assuming that `A` is originated from `LayoutLeft` or `LayoutRight`).